lms/djangoapps/gdpr_user_retirement/views.py

```diff
@@ -0,0 +1,83 @@
+"""
+An API for retiring user accounts.
+"""
+import logging
+
+from edx_rest_framework_extensions.auth.jwt.authentication import JwtAuthentication
+from django.contrib.auth import get_user_model
+from django.db import transaction
+from rest_framework import permissions, status
+from rest_framework.response import Response
+from rest_framework.views import APIView
+from six import text_type
+from social_django.models import UserSocialAuth
+from student.models import AccountRecovery, Registration, get_retired_email_by_email
+from openedx.core.djangolib.oauth2_retirement_utils import retire_dot_oauth2_models
+from openedx.core.djangoapps.user_api.models import UserRetirementStatus
+from openedx.core.djangoapps.user_api.accounts.permissions import CanRetireUser
+
+log = logging.getLogger(__name__)
+
+
+class GDPRUsersRetirementView(APIView):
+    """
+    **Use Case**
+
+        Implementation for GDPR User Retirement API. Creates a retirement request
+        for one or more users.
+
+    **Example Request**
+
+        POST /v1/accounts/gdpr_retire_users {
+            "usernames": "test_user1, test_user2"
+        }
+
+        **POST Parameters**
+
+          A POST request can include the following parameter.
+
+          * usernames: Comma separated strings of usernames that should be retired.
+    """
+    authentication_classes = (JwtAuthentication, )
+    permission_classes = (permissions.IsAuthenticated, CanRetireUser)
+
+    def post(self, request, **kwargs):
+        """
+        Initiates the GDPR retirement process for the given users.
+        """
+        request_body = request.data.get('usernames')
+        if request_body:
+            usernames_to_retire = [each_username.strip() for each_username in request_body.split(',')]
+        else:
+            usernames_to_retire = []
+        User = get_user_model()
+        for username in usernames_to_retire:
+            try:
+                user_to_retire = User.objects.get(username=username)
+                with transaction.atomic():
+                    # Add user to retirement queue.
+                    UserRetirementStatus.create_retirement(user_to_retire)
+                    # Unlink LMS social auth accounts
+                    UserSocialAuth.objects.filter(user_id=request.user.id).delete()
+                    # Change LMS password & email
+                    user_to_retire.email = get_retired_email_by_email(user_to_retire.email)
+                    user_to_retire.set_unusable_password()
+                    user_to_retire.save()
+
+                    # Remove the activation keys sent by email to the user for account activation.
+                    Registration.objects.filter(user=user_to_retire).delete()
+
+                    # Delete OAuth tokens associated with the user.
+                    retire_dot_oauth2_models(user_to_retire)
+                    AccountRecovery.retire_recovery_email(request.user.id)
```

Comment:
Does the above logic exist somewhere else? How does this flow happen today, without this API? Seems like these steps should be defined in a common place.

Is it possible / reasonable to make it so that the "retire my account" button in a user's own user settings uses this API too, to avoid duplication? What API does that use btw?

Reply:
The existing process is as follows:
1. The user clicks on "Delete my account" button on their account settings page. The auth tokens, email id and passwords are reset and a User Retirement request is created.
2. A separate request needs to be made (usually after some cool-off time) which will process the User Retirement request and perform the actual PII removal from all locations.

Currently, there is no other way to create a User Retirement request via API (It can only be created when the user clicks "Delete my account" or via admin panel). This PR is basically introducing an API to perform step 1 above.

Reply:
Done :+1: